### DISCUSSION TITLE
Missing Documentation? peripheral presets defined how?

### DISCUSSION THREAD

**Discussion by tablatronix** [arendst/Tasmota#12165] [Q&A] [answered]
Peripheral presets
button_/switch_ ?

I cannot find any documentation on these, and could not glean from source either so far.

Clearly these are some kind of presets, inverted, pullup, touch control ?? But I cannot for the life of me find where they are defined or how. Or what they do./difference

![Screen Shot 2021-05-22 at 9 10 10 AM](https://user-images.githubusercontent.com/807787/119237973-60e84f00-bb05-11eb-828e-efbaed27b844.png)

**Comment by sfromis** [comment, marked as answer]
I already answered you on discord.
https://tasmota.github.io/docs/Components/

You have to know your hardware to select details.

**Reply by tablatronix** [reply]
you linked to a doc page, which I of course already read.
It doesn't answer my question of what these do and are defined as under generic module type, not sure what hardware has to do with it.

**Reply by sfromis** [reply]
If you know your hardware, you know when you need things like inverted or no pullup. And you know when you try to use touch control. Do not expect Tasmota docs to be a course in getting to know how hardware works.

**Reply by tablatronix** [reply]
I am asking what these presets are defined as ,what they mean.. Not when to use them or why.
what docs and I missing that say what postfix _n, _i, _tc mean or how/where each one is defined in source

**Reply by sfromis** [reply]
Well, if you know your hardware, you know what inverted means. Very logically, inverted means inverted between on/off.
This page does have a few extra explanation. https://tasmota.github.io/docs/Buttons-schematics/

**Reply by tablatronix** [reply]
ah that diagram does have the explanations, they should be in the docs in text `button_n(external pullup)` that is not described anywhere that I can find.

I mean literally the preset postfix codes in tasmota itself. what is `_i` what is `_n` what is the definition behind these preset dropdowns. The generic seems to have all* so they come from somewhere.

I am not asking about real world stuff, I know hardware and software very well.
